spec.md

```diff
@@ -0,0 +1,954 @@
+- This operation MUST be idempotent.
+
+- The `CreateMachineResponse` returned by this method is expected to return
+    - `MachineID` that uniquely identifys the VM at the provider. This is expected n the  matches with the node.Spec.ProviderID on the node object.
+    - `NodeName` that is the expected name of the machine when it joins the cluster. It must match with the node name.
+
+```protobuf
+message CreateMachineRequest {
+    // Name is the name of the machine to be created by driver.
+    // This field is REQUIRED.
+    string Name = 1;
+
+    // ProviderSpec is the configuration needed to create a machine in bytes.
+    // Driver should parse this raw data into pre-defined spec in their respective projects.
+    // This field is REQUIRED.
+    bytes ProviderSpec = 2;
+
+    // Secrets is the map containing necessary credentials for cloud-provider to create the machine.
+    // This field is OPTIONAL.
+    map<string, bytes> Secrets = 3 [(cmi_secret) = true];
+}
+
+message CreateMachineResponse {
+    // MachineID is the unique identification of the VM at the cloud provider.
+    // This could be the same/different from req.Name.
+    // MachineID typically matches with the node.Spec.ProviderID on the node object.
+    // Eg: gce://project-name/region/vm-machineID
+    // This field is REQUIRED.
+    string MachineID = 1;
+
+    // NodeName is the name of the node-object registered to kubernetes.
+    // This field is REQUIRED.
+    string NodeName = 2;
+}
+```
+
+##### CreateMachine Errors
+
+If the plugin is unable to complete the CreateMachine call successfully, it MUST return a non-ok gRPC code in the gRPC status.
+If the conditions defined below are encountered, the plugin MUST return the specified gRPC error code.
+The CMI-Client MUST implement the specified error recovery behavior when it encounters the gRPC error code.
+
+| gRPC Code | Condition | Description | Recovery Behavior | Auto Retry Required |
+|-----------|-----------|-------------|-------------------|------------|
+| 0 OK | Successful | The call was successful in creating/adopting a VM that matches supplied creation request. The `CreateMachineResponse` is returned with desired values |  | N |
+| 1 CANCELED | Cancelled | Call was cancelled. Perform any pending clean-up tasks and return the call |  | N |
+| 2 UNKNOWN | Something went wrong | Not enough information on what went wrong | Retry operation after sometime | Y |
+| 3 INVALID_ARGUMENT | Re-check supplied parameters | Re-check the supplied `Name` and `ProviderSpec`. Make sure all parameters are in permitted range of values. Exact issue to be given in `.message` | Update providerSpec to fix issues. | N |
+| 4 DEADLINE_EXCEEDED | Timeout | The call processing exceeded supplied deadline | Retry operation after sometime | Y |
```

Comment:
how will one come out of it if the supplied deadline is managed by cmi-client and not the provider as retry may never succeed if the provider takes more time then what is set as default timeout ? related to the comment I made in an earlier section.

Reply:
Okay, let's take an example of a machine-creation taking up to 20mins on a provider XYZ and machine creation timeout set to 5mins by `CMI-Client`. Does it make sense for the plugin to partially process creation request, and on deadline exceed continue the creation in the next creation request? So in this case, the machine creation may take 4-5 creation calls all of whom return non-0 error with retry except for the last call. 

Will this make sense to the plugin implementor?

Reply:
@prashanth26 This is exactly what makes sense IMO, and how we've already implemented it.